### PR TITLE
feat: proof upgrade testing

### PR DESCRIPTION
**Pull request by cryptonemo**
This PR is NOT intended to be merged, as it contains incompatible proofs behaviour.  At the moment, this branch is for testing.

### CODE REVIEW

File: filecoin-proofs/src/api/post.rs [hunk 1]
```diff
@@ -604,18 +595,18 @@ pub fn verify_winning_post<Tree: 'static + MerkleTreeTrait>(
...
-            pub_sectors.push(fallback::PublicSector {
+            Ok(fallback::PublicSector {
                 id: *sector_id,
                 comm_r,
-            });
-        }
-    }
+            })
+        })
+        .collect::<Result<_>>()?;
```
Comment (porcuquine) [inline filecoin-proofs/src/api/post.rs]: I think the problem with this change is that it assumed each sector could only appear once in the set of challenged sectors. But in the case of Winning PoSt, there is just one sector, yet it is challenged repeatedly (each time with only a single challenge).

Yes, this is strange and confusing. Yes, the pervasive naming makes this almost impossible to understand.

File: filecoin-proofs/src/api/post.rs [hunk 2]
```diff
@@ -452,6 +452,7 @@ pub fn generate_fallback_sector_challenges<Tree: 'static + MerkleTreeTrait>(
     randomness: &ChallengeSeed,
     pub_sectors: &[SectorId],
     _prover_id: ProverId,
+    is_winning: bool,
```
Comment (porcuquine) [inline filecoin-proofs/src/api/post.rs:455]: If we want to avoid completely splitting Window and Winning PoSt, we need to be able to signal which we are using. We could try to 'detect' somehow based on the actual values, but that would be very brittle and a bad path to follow.

File: filecoin-proofs/src/api/post.rs [hunk 3]
```diff
@@ -595,18 +598,18 @@ pub fn verify_winning_post<Tree: 'static + MerkleTreeTrait>(
...
-            Ok(fallback::PublicSector {
+            pub_sectors.push(fallback::PublicSector {
                 id: *sector_id,
                 comm_r,
-            })
-        })
-        .collect::<Result<_>>()?;
+            });
+        }
+    }
```
Comment (porcuquine) [inline filecoin-proofs/src/api/post.rs]: We need to roll this back. Winning PoSt will not actually change its behavior. We need the challenges of Winning PoSt to be sensitive to order, because otherwise they would all be the same. This is because a single challenge appears repeatedly at different positions within the list of challenged sectors. Each instance needs fresh challenges.

File: storage-proofs/post/src/fallback/compound.rs [hunk 4]
```diff
@@ -68,7 +68,11 @@ impl<'a, Tree: 'static + MerkleTreeTrait>
                     &pub_params,
                     pub_inputs.randomness,
                     sector.id.into(),
-                    challenge_index as u64,
+                    if pub_params.is_winning {
+                        i
+                    } else {
+                        challenge_index
+                    } as u64,
```
Comment (porcuquine) [inline storage-proofs/post/src/fallback/compound.rs:75]: This works, but it's a hack which now depends on each Winning PoSt sector being challenged only once. To be more correct, the `true` branch of this conditional should fully restore previous behavior. In practice, this won't matter with current parameterization, but we shouldn't depend on that. Certainly we would need at least an assertion if we intended to. That might actually be preferable. We have an enormous amount of complexity in place to support a generality I don't think we need. If we can close off some possibilities for the sake of simpler, clearer, code that might be worth doing.

File: storage-proofs/post/src/fallback/compound.rs [hunk 5]
```diff
@@ -234,6 +238,7 @@ mod tests {
                 sector_size: sector_size as u64,
                 challenge_count,
                 sector_count,
+                is_winning: false,
```
Comment (porcuquine) [inline storage-proofs/post/src/fallback/compound.rs:241]: For the sake of getting this working quickly, I ended up specializing all the tests to Window PoSt (`is_winning: false`). To be thorough, they should probably be extended to exercise both code paths.

File: filecoin-proofs/tests/api.rs [hunk 6]
```diff
@@ -375,8 +373,11 @@ fn window_post<Tree: 'static + MerkleTreeTrait>(
     /////////////////////////////////////////////
     // The following methods of proof generation are functionally equivalent:
     // 1)
-    //let proof = generate_window_post::<Tree>(&config, &randomness, &priv_replicas, prover_id)?;
-    //
+    let proof = generate_window_post::<Tree>(&config, &randomness, &priv_replicas, prover_id)?;
```
Comment (porcuquine) [inline filecoin-proofs/tests/api.rs:376]: Commit description is misleading (mentioning for anyone else trying to make sense of it). It's Window PoSt that is affected.

Comment (porcuquine) [reply filecoin-proofs/tests/api.rs:376]: No need.

File: storage-proofs/post/src/fallback/vanilla.rs [hunk 7]
```diff
@@ -380,16 +383,28 @@ impl<'a, Tree: 'a + MerkleTreeTrait> ProofScheme<'a> for FallbackPoSt<'a, Tree>
                     Tree::Arity::to_usize(),
                 );
 
+                let num_challenges = if pub_params.is_winning {
+                    pub_params.sector_count
```
Comment (porcuquine) [inline storage-proofs/post/src/fallback/vanilla.rs:387]: I think this depends on `pub_params.challenge_count` being equal to 1. Since that's not structurally guaranteed (but will be the case in extant WindowPoSt), I think it would be good to also assert it here. I'd rather be safe than sorry, at least until we can clean this up more broadly.

File: storage-proofs/post/src/fallback/compound.rs [hunk 8]
```diff
@@ -65,10 +64,16 @@ impl<'a, Tree: 'static + MerkleTreeTrait>
...
             for n in 0..pub_params.challenge_count {
-                let challenge_index = ((partition_index * pub_params.sector_count + i)
-                    * pub_params.challenge_count
-                    + n) as u64;
-                let challenged_leaf_start = fallback::generate_leaf_challenge(
+                let challenge_index = if pub_params.is_winning {
+                    // Note that this generality is perhaps over-complicated and unnecessary
+                    // with the current parameterization.  To avoid complexity, the
+                    // challenge_index could be set to 'i' here.
+                    (partition_index * pub_params.sector_count + i) * pub_params.challenge_count + n
```
Comment (porcuquine) [inline storage-proofs/post/src/fallback/compound.rs:71]: Can we guard the simplifying assumption with an assertion then revert to the simple expression? That works for me.

File: storage-proofs/post/src/fallback/vanilla.rs [hunk 9]
```diff
@@ -427,11 +434,25 @@ impl<'a, Tree: 'a + MerkleTreeTrait> ProofScheme<'a> for FallbackPoSt<'a, Tree>
...
                 }
 
-                proofs.push(SectorProof {
-                    inclusion_proofs,
-                    comm_c: priv_sector.comm_c,
-                    comm_r_last: priv_sector.comm_r_last,
-                });
+                // Winning PoSt and Window PoSt have different proof shapes regarding inclusion
+                // and sector proofs, so we are careful to partition them appropriately.
+                if pub_params.is_winning {
```
Comment (porcuquine) [inline storage-proofs/post/src/fallback/vanilla.rs:439]: My experience of reading this whole section now is that it feels obfuscated and is very hard to follow. I am starting to suspect it would be better to just select on type (`is_winning`) then clearly and concisely perform the appropriate work in each branch. Treating these as 'same but different' with all the conditionals, hidden assumptions, and gymnastics doesn't seem worth it. If nothing else, it makes the code hard to follow — and it also makes it hard to reason about well enough to gain confidence. I think that goal might be more achievable with the alternate structure I just proposed. What do you think?

Comment (cryptonemo) [reply storage-proofs/post/src/fallback/vanilla.rs:439]: It's a good idea for sure -- but the reason I'd hold off is because we already do something like this here (arguably the 'improved' version, or at least inspired by the idea): https://github.com/filecoin-project/rust-fil-proofs/blob/master/filecoin-proofs/src/api/post.rs#L672 and ideally, as part of a larger re-factor, we can share code across as needed better (rather than duplicating).

Comment (porcuquine) [reply storage-proofs/post/src/fallback/vanilla.rs:445]: Do you want to do that now then? My concern is trying to avoid accidentally releasing broken code. I'm not saying I believe this code is necessarily broken, but if we let the mandate of an upgrade complicate the code base, then we're at risk of making it very hard to continue maintaining in case further change becomes required.

I would personally prefer to maximize simplicity, even at the cost of some 'duplication' which can be refactored away later. Or do the refactor now, if not too difficult. The only reason the code isn't duplicated now is because we have two *different* methods, one of which is very hard to comprehend.

**Comment by porcuquine** [comment]
@cryptonemo I think my previous commit is ready for review. This is about all the refactoring that feels needed for now (apart from whatever is logically implied that I might have overlooked).

File: fil-proofs-tooling/src/bin/benchy/prodbench.rs [hunk 10]
```diff
@@ -313,18 +317,18 @@ fn generate_params(i: &ProdbenchInputs) {
         sector_size,
         partitions,
         porep_id: dummy_porep_id,
+        api_version: FIXED_API_VERSION,
```
Comment (dignifiedquire) [inline fil-proofs-tooling/src/bin/benchy/prodbench.rs:320]: this should probably a parameter of the tooling

File: fil-proofs-tooling/src/bin/benchy/window_post.rs [hunk 11]
```diff
@@ -514,6 +516,7 @@ pub fn run_window_post_bench<Tree: 'static + MerkleTreeTrait>(
             .expect("unknown sector size"),
         typ: PoStType::Window,
         priority: true,
+        api_version: APIVersion::V1_1,
```
Comment (dignifiedquire) [inline fil-proofs-tooling/src/bin/benchy/window_post.rs:519]: should probably a parameter

File: storage-proofs/core/src/api_version.rs [hunk 12]
```diff
@@ -1,5 +1,21 @@
...
+
 #[derive(Debug, Copy, Clone)]
 pub enum APIVersion {
     V1_0,
     V1_1,
 }
+
+impl FromStr for APIVersion {
+    type Err = anyhow::Error;
+    fn from_str(api_version: &str) -> Result<Self> {
```
Comment (dignifiedquire) [inline storage-proofs/core/src/api_version.rs:12]: Lets associate a `Semver` version with each of the enum and use that for parsing using: https://docs.rs/semver/

File: storage-proofs/core/src/api_version.rs [hunk 13]
```diff
@@ -1,5 +1,21 @@
+use anyhow::{self, Result};
+use std::str::FromStr;
+
 #[derive(Debug, Copy, Clone)]
 pub enum APIVersion {
```
Comment (dignifiedquire) [inline storage-proofs/core/src/api_version.rs:5]: Lets call this `ApiVersion`, it is a little bit more consistent with naming conventions in rust

File: storage-proofs/core/src/api_version.rs (hunk 12 above)
Comment (cryptonemo) [reply storage-proofs/core/src/api_version.rs:12]: For this, do we want `ApiVersion::{V1_0_0, V1_1_0}` for the constants also?

Comment (dignifiedquire) [reply storage-proofs/core/src/api_version.rs:12]: good question, I am not sure tbh
